packages/auth/amplify_auth_cognito/lib/src/state/event/credential_store_event.dart

```diff
@@ -0,0 +1,290 @@
+/// Initiates migration of a legacy credential store (pre-vNext) to the updated
+/// structure.
+/// {@endtemplate}
+class CredentialStoreMigrateLegacyCredentialStore extends CredentialStoreEvent {
+  /// {@macro amplify_auth_cognito.migrate_legacy_credential_store}
+  const CredentialStoreMigrateLegacyCredentialStore() : super._();
+
+  @override
+  CredentialStoreEventType get type =>
+      CredentialStoreEventType.migrateLegacyCredentialStore;
+
+  @override
+  List<Object?> get props => [type];
+
+  @override
+  String? checkPrecondition(CredentialStoreState currentState) {
+    if (currentState.type !=
+        CredentialStoreStateType.loadingStoredCredentials) {
+      return 'Credential store cannot be migrated in current state';
+    }
+    return null;
+  }
+}
+
+/// {@template amplify_auth_cognito.store_credentials}
+/// Initiates storing of user credentials.
+/// {@endtemplate}
+class CredentialStoreStoreCredentials extends CredentialStoreEvent {
+  /// {@macro amplify_auth_cognito.store_credentials}
+  const CredentialStoreStoreCredentials({
+    this.identityId,
+    this.awsCredentials,
+    this.userPoolTokens,
+    this.deviceSecrets,
+  }) : super._();
+
+  /// AWS Identity ID
+  final String? identityId;
+
+  /// AWS Identity Pool credentials
+  final AWSCredentials? awsCredentials;
+
+  /// Cognito User Pool tokens
+  final CognitoUserPoolTokens? userPoolTokens;
+
+  /// Registered device secrets
+  final CognitoDeviceSecrets? deviceSecrets;
+
+  @override
+  CredentialStoreEventType get type =>
+      CredentialStoreEventType.storeCredentials;
+
+  @override
+  List<Object?> get props => [
+        type,
+        identityId,
+        awsCredentials,
+        userPoolTokens,
+        deviceSecrets,
+      ];
+
+  @override
+  String? checkPrecondition(CredentialStoreState currentState) {
+    if (currentState.type == CredentialStoreStateType.notConfigured) {
+      return 'Credential store is not configured';
+    }
+    if (currentState.type == CredentialStoreStateType.failure) {
+      return 'Credential store has error. Re-load before continuing.';
+    }
+    if (currentState.type != CredentialStoreStateType.success) {
+      return 'Credential store is busy';
+    }
+    return null;
+  }
+}
+
+/// {@template amplify_auth_cognito.clear_credentials}
+/// Initiates clearing of the credential store.
+/// {@endtemplate}
+class CredentialStoreClearCredentials extends CredentialStoreEvent {
+  /// {@macro amplify_auth_cognito.clear_credentials}
+  const CredentialStoreClearCredentials([this.keys = const []]) : super._();
+
+  /// When set, only these keys will be cleared from the store. Otherwise,
+  /// all keys are cleared.
+  final Iterable<String> keys;
+
+  @override
+  CredentialStoreEventType get type =>
+      CredentialStoreEventType.clearCredentials;
+
+  @override
+  List<Object?> get props => [type, keys];
+
+  @override
+  String? checkPrecondition(CredentialStoreState currentState) {
```

Comment:
nit: this looks identical to same method in `CredentialStoreStoreCredentials`. Could it be DRYed either by moving to helper function? Or maybe a shared subclass?

Reply:
Yeah, it probably could be. Eventually, all of these event/state types will be auto-generated by Preamp.